### PR TITLE
Add runtime signature into deps.json

### PR DESCRIPTION
**Pull request by pakrym**
@anurse @schellap

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/2016)

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 1]
```diff
@@ -48,16 +50,38 @@ public DependencyContextBuilder(string referenceAssembliesPath)
...
+        private static string GenerateRuntimeSignature(IEnumerable<LibraryExport> runtimeExports)
+        {
+            var sha1 = SHA1.Create();
+            var builder = new StringBuilder();
+            var packages = runtimeExports
+                .Where(libraryExport => libraryExport.Library.Identity.Type == LibraryType.Package);
+            foreach (var libraryExport in packages)
+            {
+                builder.Append(libraryExport.Library.Identity.Name);
+                builder.Append(libraryExport.Library.Identity.Version);
```
Comment (anurse) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: Could you add a separator between each individual library, just to be safe? A newline would be sufficient, so just changing this to `AppendLine`

**Comment by anurse** [comment]
:shipit: on the managed code. Native code looks fine but wait for @schellap 's sign off on that. Also wait for whatever craziness is going on with the builds :)

File: src/corehost/cli/deps_format.cpp [hunk 2]
```diff
@@ -318,8 +318,11 @@ bool deps_json_t::load(bool portable, const pal::string_t& deps_path, const rid_
     {
         const auto json = json_value::parse(file);
 
-        const auto& runtime_target = json.at(_X("runtimeTarget"));
-        const pal::string_t& name = runtime_target.as_string();
+		const auto& runtime_target = json.at(_X("runtimeTarget"));
+
+        const pal::string_t& name = (runtime_target.type() == json_value::value_type::String)?
+			runtime_target.as_string():
+			runtime_target.at(_X("name")).as_string();
```
Comment (schellap) [inline src/corehost/cli/deps_format.cpp]: NIT: Whitespace issues.

`runtime_target.is_string() ?`

is simpler. Otherwise, .cpp looks good to me.

**Comment by anurse** [comment]
FYI https://github.com/dotnet/cli/pull/2019 should fix the build issues

File: test/Microsoft.Extensions.DependencyModel.Tests/DependencyContextBuilderTests.cs [hunk 3]
```diff
@@ -291,6 +291,19 @@ public void SkipsBuildDependencies()
...
+        [Fact]
+        public void GeneratesRuntimeSignatureOutOfPackageNamesAndVersions()
+        {
+            var context = Build(runtimeExports: new[]
+            {
+                Export(PackageDescription("Pack.Age", new NuGetVersion(1, 2, 3))),
+                Export(PackageDescription("Pack.Age", new NuGetVersion(1, 2, 3))),
+            });
+
+            var lib = context.Target.RuntimeSignature.Should().Be("85aa350d4f334de42348cdfc5e09f4ae9a328eeb");
```
Comment (davidfowl) [inline test/Microsoft.Extensions.DependencyModel.Tests/DependencyContextBuilderTests.cs]: lib?

**Comment by schellap** [comment]
Just wanted to make sure you are adding signature of serviceable packages... a small optimization.

**Comment by schellap** [comment]
@pakrym, @anurse, @davidfowl, any thoughts about just doing it for serviceable case or be generic.

**Comment by anurse** [comment]
We just did it for all packages. It may mean occasionally over-breadcrumbing but I think that's OK.

**Pull request merged**